### PR TITLE
Add a way to register and get Embeddings

### PR DESCRIPTION
**Pull request by joelhoisko**
- New inline `withEmbeddings()` builder.
- New `@embedding('guid')`, `@compare()`, `@on()` and `@deleteMethod()` decorators for declaring an embedding class
	- Names for `@deleteMethod()` are welcome, can't use `delete` as it's a reserved keyword. Other possibilites could be `@remove` or the more powerful `@destroy` or even `@obliterate`
	- Embeddings need their own special `@on()` method to capture their slightly different method signatures. It's easy to accidentally use the `@on()` decorator from projections instead here. I tried to make it very cool and generic for a couple of hours but failed miserably, maybe somebody smarter can figure it out (problem being that you can't figure out type aliases on runtime 🙂)
- Get embddings with `get()`, `getAll()` or just keys with `getKeys()` on `client.embeddings`
- `events.processing.Internal.EventProcessor` now extends the new `ClientProcessor` from `services`. I don't think that the way I split it out was that elegant, but it works. Opinions are welcome.
- `OnMethodBuilder` makes it easier to build the different callbacks for the `@on()` decorators

There are some problems here, like the embeddings still using the same `IProjectionAssociations`, that has projection specific exceptions.

**Examples:**

Inline:
```typescript
.withEmbeddings(builder => {
    builder.register(DishCounter);
    builder.createEmbedding('0767bc04-bc03-40b8-a0be-5f6c6130f68b')
        .forReadModel(Chef)
        .compare((oldState, newState, embeddingContext) => {
            return oldState.dishes
                .filter((dish: string) => !newState.dishes.includes(dish))
                .map((missingDish: string) => new DishPrepared(missingDish, 'default chef lol'))
        })
        .deleteMethod((currentState, embeddingContext) => {
            return new ChefFired(currentState.name);
        })
        .on(DishPrepared, _ => _.keyFromProperty('Chef'), (chef, event, projectionContext) => {
            chef.name = event.Chef;
            if (!chef.dishes.includes(event.Dish)) chef.dishes.push(event.Dish);
            return chef;
        });
    })
```

Class:
```typescript
@embedding('98f9db66-b6ca-4e5f-9fc3-638626c9ecfa')
export class DishCounter {
    numberOfTimesPrepared: number = 0;

    @compare()
    compare(receivedState: DishCounter, embeddingContext: EmbeddingContext) {
        if (receivedState.numberOfTimesPrepared > this.numberOfTimesPrepared) {
            return new DishPrepared(embeddingContext.key.value, 'default chef');
        }
    }

    @deleteMethod()
    remove(embeddingContext: EmbeddingContext) {
        return new DishRemoved(embeddingContext.key.value);
    }

    @on(DishPrepared, _ => _.keyFromProperty('Dish'))
    on(event: DishPrepared, projectionContext: ProjectionContext) {
        this.numberOfTimesPrepared ++;
    }
}
```

Get embeddings:
```typescript
for (const [dish, { state: counter }] of await client.embeddings.forTenant(TenantId.development).getAll(DishCounter)) {
    console.log(`The kitchen has prepared ${dish} ${counter.numberOfTimesPrepared} times`);
}

const chef = await client.embeddings.forTenant(TenantId.development).get<Chef>(Chef, 'Mrs. Tex Mex');
console.log(`${chef.key} has prepared ${chef.state.dishes}`);

const dishes = await client.embeddings.forTenant(TenantId.development).getKeys(DishCounter);
console.log(`Got dem dish keys: ${dishes}`);
```

### CODE REVIEW

File: Source/embeddings/Store/EmbeddingStore.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+import { reactiveUnary } from '@dolittle/sdk.services';
+import { Constructor } from '@dolittle/types';
+import { map } from 'rxjs/operators';
+import { Logger } from 'winston';
+import { EmbeddingId, FailedToGetEmbeddingKeys } from '..';
+import { FailedToGetEmbedding } from './FailedToGetEmbedding';
+import { FailedToGetEmbeddingState } from './FailedToGetEmbeddingState';
+import { IEmbeddingStore } from './IEmbeddingStore';
+
+/** @inheritdoc */
```
Comment (woksin) [inline Source/embeddings/Store/EmbeddingStore.ts:20]:
```suggestion
/**
 * Represents an implementation of {link IEmbeddingStore}
 */
```

File: Source/embeddings/Store/FailedToGetEmbedding.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import { ScopeId } from '@dolittle/sdk.events';
+import { Failure } from '@dolittle/sdk.protobuf';
+import { Key } from '@dolittle/sdk.projections';
+import { EmbeddingId } from '..';
+
+/**
+ * An exception that gets thrown when an embedding fails to be getted.
+ */
+export class FailedToGetEmbedding extends Exception {
+    constructor(embedding: EmbeddingId, key: Key | undefined, failure: Failure) {
```
Comment (woksin) [inline Source/embeddings/Store/FailedToGetEmbedding.ts:14]:
```suggestion
    /**
     * Initialises an instance of {@link FailedToGetEmbedding}.
     * @param {EmbeddingId} embedding The embedding identifier.
     * @param {Key | undefined} key The optional key.
     * @param {Failure} failure The failure.
     */ 
    constructor(embedding: EmbeddingId, key: Key | undefined, failure: Failure) {
```

File: Source/embeddings/Store/FailedToGetEmbedding.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        if (key) {
+            super(`Failed to get embedding ${embedding} with key ${key} due to failure ${failure.id} with reason: ${failure.reason}`);
+        } else {
+            super(`Failed to get embedding ${embedding} due to failure ${failure.id} with reason: ${failure.reason}`);
+        }
```
Comment (woksin) [inline Source/embeddings/Store/FailedToGetEmbedding.ts:19]: We could do this instead?
```suggestion
        super(`Failed to get embedding ${embedding} ${key ? `with key ${key}  `: ""}due to failure ${failure.id} with reason: ${failure.reason}`);
```

File: Source/embeddings/Store/FailedToGetEmbeddingKeys.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+import { Exception } from '@dolittle/rudiments';
+import { Failure } from '@dolittle/sdk.protobuf';
+import { EmbeddingId } from '..';
+
+/**
+ * An exception that gets thrown when an embedding fails to be getted.
+ */
+export class FailedToGetEmbeddingKeys extends Exception {
+    constructor(embedding: EmbeddingId, failure: Failure) {
```
Comment (woksin) [inline Source/embeddings/Store/FailedToGetEmbeddingKeys.ts:12]:
```suggestion
    /**
     * Initialises an instance of {@link FailedToGetEmbeddingKeys}.
     * @param {EmbeddingId} embedding The embedding identifier.
     * @param {Failure} failure The failure.
     */
    constructor(embedding: EmbeddingId, failure: Failure) {
```

File: Source/embeddings/Store/FailedToGetEmbeddingState.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+import { Exception } from '@dolittle/rudiments';
+import { Key } from '@dolittle/sdk.projections';
+import { EmbeddingId } from '..';
+
+
+export class FailedToGetEmbeddingState extends Exception {
```
Comment (woksin) [inline Source/embeddings/Store/FailedToGetEmbeddingState.ts:9]:
```suggestion
/**
 * Exception that gets thrown when embedding state fails to be getted.
 */
export class FailedToGetEmbeddingState extends Exception {
```

File: Source/embeddings/Store/FailedToGetEmbeddingState.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    constructor(embedding: EmbeddingId, key: Key) {
```
Comment (woksin) [inline Source/embeddings/Store/FailedToGetEmbeddingState.ts:10]:
```suggestion
    /**
     * Initialises an instance of {@link FailedToGetEmbeddingState}.
     * @param {EmbeddingId} embedding The embedding identifier.
     * @param {Key} key The projection key.
     */
    constructor(embedding: EmbeddingId, key: Key) {
```

File: Source/embeddings/index.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+export { EmbeddingCompareCallback } from './EmbeddingCompareCallback';
+export { EmbeddingContext } from './EmbeddingContext';
+export { EmbeddingDeleteCallback } from './EmbeddingDeleteCallback';
+export { EmbeddingId } from './EmbeddingId';
+export { EmbeddingProjectCallback } from './EmbeddingProjectCallback';
+export { EmbeddingProjectContext } from './EmbeddingProjectContext';
+export { Embeddings } from './Embeddings';
+export { IEmbedding } from './IEmbedding';
+export { IEmbeddings } from './IEmbeddings';
+export * as internal from './Internal';
```
Comment (woksin) [inline Source/embeddings/index.ts:43]: Can't we just export things from here as we do above with { }?

File: Source/events/EventConverters.ts [hunk 8]
```diff
@@ -46,6 +46,21 @@ export class EventConverters {
...
+     * @param {boolean} isPublic Whether or not it is a public event
+     * @returns {UncommittedEvent} Constructed uncommitted event.
+     */
+    static getUncommittedEmbeddingEventFrom(event: any, eventType: EventType, isPublic: boolean): PbUncommittedEvent {
+        const uncommittedEvent = new PbUncommittedEvent();
+        uncommittedEvent.setArtifact(eventTypes.toProtobuf(eventType));
+        uncommittedEvent.setPublic(isPublic);
+        uncommittedEvent.setContent(JSON.stringify(event));
+        return uncommittedEvent;
+    }
```
Comment (woksin) [inline Source/events/EventConverters.ts:62]: I don't really see why we need this. If so it does not really belong here in this class

File: Source/projections/Builder/OnDecoratedProjectionMethods.ts [hunk 9]
```diff
@@ -2,30 +2,28 @@
...
 import { KeySelector } from '..';
-
-import { OnDecoratedMethod  } from './OnDecoratedMethod';
-import { ProjectionClassOnMethod  } from './ProjectionClassOnMethod';
+import { OnDecoratedProjectionMethod } from './OnDecoratedProjectionMethod';
+import { ProjectionClassOnMethod } from './ProjectionClassOnMethod';
 
 /**
- * Defines the system that knows about all the methods decorated with the on decorator.
+ * Defines the system that knows about all the methods decorated with the on decorator on a projection.
```
Comment (woksin) [inline Source/projections/Builder/OnDecoratedProjectionMethods.ts:12]:
```suggestion
 * Represents the system that knows about all the methods decorated with the on decorator on a projection.
```

File: Source/projections/Builder/OnMethodBuilder.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+import { Guid } from '@dolittle/rudiments';
+import { EventType, EventTypeId, EventTypeMap, Generation, IEventTypes } from '@dolittle/sdk.events';
+import { Constructor } from '@dolittle/types';
+import { KeySelector, KeySelectorBuilder } from '..';
+import { KeySelectorBuilderCallback } from './KeySelectorBuilderCallback';
+import { OnMethodSpecification } from './OnMethodSpecification';
+import { TypeOrEventType } from './TypeOrEventType';
+
+/**
+ * Defines a class for building and getting the inline on() methods for inline projections.
```
Comment (woksin) [inline Source/projections/Builder/OnMethodBuilder.ts:13]:
```suggestion
 * Represents a builder for building and getting the inline on() methods for inline projections.
```

File: Source/projections/Builder/ProjectionClassBuilder.ts [hunk 11]
```diff
@@ -1,27 +1,23 @@
...
-import { OnDecoratedMethod } from './OnDecoratedMethod';
-import { OnDecoratedMethods } from './OnDecoratedMethods';
-import { projection as projectionDecorator } from './projectionDecorator';
+import { OnDecoratedProjectionMethod } from './OnDecoratedProjectionMethod';
+import { OnDecoratedProjectionMethods } from './OnDecoratedProjectionMethods';
 import { ProjectionDecoratedTypes } from './ProjectionDecoratedTypes';
+import { projection as projectionDecorator } from './projectionDecorator';
+
 
 export class ProjectionClassBuilder<T> implements ICanBuildAndRegisterAProjection {
```
Comment (woksin) [inline Source/projections/Builder/ProjectionClassBuilder.ts:22]:
```suggestion
/**
 * Represents a builder for building a projection class.
 */
export class ProjectionClassBuilder<T> implements ICanBuildAndRegisterAProjection {
```

File: Source/projections/index.ts [hunk 12]
```diff
@@ -1,63 +1,62 @@
...
     ProjectionsBuilderCallback,
     ReadModelAlreadyDefinedForProjection,
+    TypeOrEventType
 } from './Builder';
-
+export { DeleteReadModelInstance } from './DeleteReadModelInstance';
+export { EventPropertyKeySelector } from './EventPropertyKeySelector';
+export { EventSelector } from './EventSelector';
+export { EventSourceIdKeySelector } from './EventSourceIdKeySelector';
+export * as internal from './Internal';
```
Comment (woksin) [inline Source/projections/index.ts:31]: Same as previous comment, shouldn't we just export things from here as we usually do

**Review by woksin — CHANGES_REQUESTED** [review]
Good stuff! Lots of code to go through, but  I did not see any glaring mistakes in the code.

File: Source/embeddings/Builder/EmbeddingsBuilder.ts [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+     * @param instance The instance to register as an event handler.
+     */
+    register<T = any>(instance: T): EmbeddingsBuilder;
+    register<T = any>(typeOrInstance: Constructor<T> | T): EmbeddingsBuilder {
+        this._embeddingBuilders.push(new EmbeddingClassBuilder<T>(typeOrInstance));
+        this._projectionAssociations.associate<T>(typeOrInstance);
+        return this;
+    }
+
+    buildAndRegister(
```
[the remark this answers is not in the document]
Comment (joelhoisko) [reply Source/embeddings/Builder/EmbeddingsBuilder.ts:54]: This class doesn't inherit anything.

File: Source/embeddings/Builder/EmbeddingBuilder.ts [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+import { EmbeddingsClient } from '@dolittle/runtime.contracts/Embeddings/Embeddings_grpc_pb';
+
+import { IEmbeddings, embedding, EmbeddingId } from '..';
+
+import { ICanBuildAndRegisterAnEmbedding } from './ICanBuildAndRegisterAnEmbedding';
+import { EmbeddingBuilderForReadModel } from './EmbeddingBuilderForReadModel';
+import { ReadModelAlreadyDefinedForEmbedding } from './ReadModelAlreadyDefinedForEmbedding';
+
+export class EmbeddingBuilder implements ICanBuildAndRegisterAnEmbedding {
```
[the remark this answers is not in the document]
Comment (joelhoisko) [reply Source/embeddings/Builder/EmbeddingBuilder.ts:22]: Fix'd the wording.

File: Source/embeddings/Builder/EmbeddingsBuilder.ts [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+import { IEmbeddings, EmbeddingId, Embeddings } from '..';
+
+import { ICanBuildAndRegisterAnEmbedding } from './ICanBuildAndRegisterAnEmbedding';
+import { EmbeddingBuilder } from './EmbeddingBuilder';
+import { EmbeddingClassBuilder } from './EmbeddingClassBuilder';
+
+export class EmbeddingsBuilder {
+    private _embeddingBuilders: ICanBuildAndRegisterAnEmbedding[] = [];
+
+    constructor(private _projectionAssociations: IProjectionAssociations) {}
```
[the remark this answers is not in the document]
Comment (joelhoisko) [reply Source/embeddings/Builder/EmbeddingsBuilder.ts:28]: Added, just removed the extra newline.

File: Source/embeddings/Builder/EmbeddingBuilderForReadModel.ts [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+            return;
+        }
+        if (this._compareMethod === undefined) {
+            logger.warn(`Failed to register embedding ${this._embeddingId}. No compare method defined.`);
+            return;
+        }
+        if (this._deleteMethod === undefined) {
+            logger.warn(`Failed to register embedding ${this._embeddingId}. No delete method defined.`);
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (joelhoisko) [reply Source/embeddings/Builder/EmbeddingBuilderForReadModel.ts:84]: I moved it out to 2 methods, `buildOnMethods()` and `canRegisterEmbedding()`

File: Source/embeddings/Embedding.ts [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+    /** @inheritdoc */
+    async on(readModel: T, event: any, eventType: EventType, context: EmbeddingProjectContext): Promise<T | DeleteReadModelInstance> {
+        if (this._eventMap.has(eventType)) {
+            const [method] = this._eventMap.get(eventType)!;
+            return method(readModel, event, context);
+        } else {
+            throw new MissingOnMethodForType(ProjectionId.from(this.embeddingId.value), eventType);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (joelhoisko) [reply Source/embeddings/Embedding.ts]: Fixd

File: Source/events/EventConverters.ts (hunk 8 above)
Comment (joelhoisko) [reply Source/events/EventConverters.ts:62]: Events committed in the `@compare` or `@deleteMethod` methods don't have an EventSourceId tied to them, as those methods just return objects. Usually when committing events, you'd do `eventStore.commit(new DishPrepared('blah'), 'eventsourceid')`, but you don't know what the eventsourceid is in the compare method and you don't need to. We can be smart about it in the runtime and set the correct eventsourceid to the events committed in the `@compare` method.

As for where it belongs, I saw that this class also has all the other Runtime <-> SDK event conversions.

File: Source/embeddings/index.ts (hunk 7 above)
Comment (joelhoisko) [reply Source/embeddings/index.ts:43]: The names could potentially clash right? Like this:
```typescript
export { EventProcessor } from './EventProcessor';
export { EventProcessor } from './Internal';
```

File: Source/events/EventConverters.ts (hunk 8 above)
Comment (woksin) [reply Source/events/EventConverters.ts:62]: I guess that it's fine for now. We'll keep it

File: Source/embeddings/Builder/EmbeddingBuilderForReadModel.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (woksin) [reply Source/embeddings/Builder/EmbeddingBuilderForReadModel.ts]: Nice

File: Source/embeddings/index.ts (hunk 7 above)
Comment (woksin) [reply Source/embeddings/index.ts:43]: I see. We could keep this for now.

**Review by woksin — APPROVED** [review]

**Pull request merged**